### PR TITLE
Slides: Design patterns

### PR DESCRIPTION
**Pull request by amanjeev**
- Adds new slide deck for basic design patterns and helpers
- Removes conversion patterns slide deck and merges those with above

### CODE REVIEW

File: presentations/design-patterns/slides.adoc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+[source,rust]
+----
+include::./4.rs[]
+----
+
+== Pattern: NewType
+
+- Use Rust type system to convey meaning to the user.
+- Especially for Types that shuold be similar to other Types.
```
Comment (justahero) [inline presentations/design-patterns/slides.adoc:74]: Typo `shuold`.

**Comment by justahero** [comment]
Thanks for putting that together, it's a topic that is frequently raised in trainings. :+1:

**Comment by justahero** [comment]
Related to the `NewType` pattern, I sometimes see a similar pattern where an external struct is extended with a useful methods.

```rust
trait VecExt {
    fn size(&self) -> usize;
}

impl<T> VecExt for Vec<T> {
    fn size(&self) -> usize {
        self.len()
    }
}

fn main() {
    let v = vec![1, 2, 3, 4, 5];
    println!("size = {}", v.size())
}
```

This adds the method `size` to the `Vec` struct, when the trait `VecExt` is imported. (ok, it's not the most useful example). This pattern surely has a name, but I fail to remember how it's called. The main difference between both patterns is, that the `NewType` pattern delegates method calls to its inner type, while this pattern adds sibling methods on the same type.

**Comment by listochkin** [comment]
I think we should add more here:

  - shadowing a variable in the same scope to narrow down its type: 
    ```Rust
    let item = returns_option();
    if let Some(item) = item { 
        // ... 
    }
    ```
    Here we can also mention "mut-unmut" - I forgot the correct term:
    ```Rust
    let mut data = ...
    change_data(data);
    let data = data; 
    // data is immutable from now on
    ```

  - ["Functional Core - Imperative Shell" principle](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell). Preferring method chaining for data transformations: `list.iter().filter().map()` vs `for { if { } }`. Also mention `.collect()` here once again, and the fact that many iterator methods are available for `Option` and `Result`, and thus you can write the same code that works on a single item and on a list of items.

  - using `#[cfg(test)]` to only compile modules in testing.

**Comment by justahero** [comment]
Regarding iterator usage patterns, maybe this is something to put into the [Iterators Again](https://ferrous-systems.github.io/teaching-material/iterators-again.html#/) deck. In particular the iterator handling of `Option` & `Result` warrant their own examples / slides. What do you think?

**Comment by amanjeev** [comment]
> Related to the `NewType` pattern, I sometimes see a similar pattern where an external struct is extended with a few useful methods.
> 
> ```rust
> trait VecExt {
>     fn size(&self) -> usize;
> }
> 
> impl<T> VecExt for Vec<T> {
>     fn size(&self) -> usize {
>         self.len()
>     }
> }
> 
> fn main() {
>     let v = vec![1, 2, 3, 4, 5];
>     println!("size = {}", v.size());
> }
> ```
> 
> This adds the method `size` to the `Vec` struct, when the trait `VecExt` is imported. (ok, it's not the most useful example). This pattern surely has a name, but I fail to remember how it's called. The main difference between both patterns is, that the `NewType` pattern encapsulates the type & calls methods on the inner type, while this pattern adds sibling methods to the same type.

I added the example, slightly changed from yours.  Also, I called it "extending Types". lol

```trait VecExt {
    fn magic_number(&self) -> usize;
}

impl<T> VecExt for Vec<T> {
    fn magic_number(&self) -> usize {
        42
    }
}

fn main() {
    let v = vec![1, 2, 3, 4, 5];
    println!("Magic Number = {}", v.magic_number());
}
```

**Comment by amanjeev** [comment]
> I think we should add more here:
> 
> * shadowing a variable in the same scope to narrow down its type:
>   ```rust
>   let item = returns_option();
>   if let Some(item) = item { 
>       // ... 
>   }
>   ```
>   Here we can also mention "mut-unmut" - I forgot the correct term:
>   ```rust
>   let mut data = ...
>   change_data(data);
>   let data = data; 
>   // data is immutable from now on
>   ```

I added these but i think the remaining two should be in their respective topic areas.  What do you think?

**Comment by amanjeev** [comment]
> Regarding iterator usage patterns, maybe this is something to put into the [Iterators Again](https://ferrous-systems.github.io/teaching-material/iterators-again.html#/) deck. In particular the iterator handling of `Option` & `Result` warrant their own examples / slides. What do you think?

yea I think it their Iterators Again slides for sure.

**Pull request merged**